src/graphics/device.js

```diff
@@ -1511,6 +1511,43 @@ Object.assign(pc, function () {
             }
         },
 
+        // In the case where a texture has more than 1 level of mip data specified, but doesn't have
+        // the full mip chain, we generate the missing levels here.
+        // This is to overcome an issue where iphone xr and xs ignores further updates to the mip data
+        // after invoking gl.generateMipmap on the texture (which was the previous method of ensuring
+        // the texture's full mip chain was complete).
+        // NOTE: this function copies texture data from the previous mip level instead of downsampling it.
+        _completePartialMipmapChain: function (texture) {
```

Comment:
One comment on this part. It feels like supporting this for *all* textures is obfuscating future user bugs. For example, if someone uploads provides a texture that only has 1024 & 512 mip levels, we will fill in all the lower levels "incorrectly". This may lead to much confusion from a user if they don't realize we do this. Arguably it would be better in that case to indicate an error and force them to fix the texture.

We *could* narrow this to only fixing the cubemaps that we know we have to fix, rather than making it fully general purpose.

Reply:
Kinda agree. But I'm going to merge this for now as it's something we can refine post-deploy.

Reply:
Another problem is if somebody sets a subset of mips that are Images rather than pixel arrays, an invalid texture is now generated, whereas before, generateMipmap would at least generate a valid texture. After the Flood actually does this. It sets an incomplete set of mips (missing the lowest levels) on a texture. Internally, the engine calls generateMipmaps and it's fine. Now, the texture is invalid and WebGL fails to set it. Despite this, I do think (now more than ever) that we should not try to clean up developers' misuse of the API. We can just enforce a rule that, if you don't supply all levels, the texture is incomplete and won't render.

Reply:
I'd rather do that than use the downsampleImage function in device.js to generate missing mipmaps!

Reply:
Given your comments above, I wonder whether it might be best to implement correct downsampling for rgba8 textures and fall back to gl.generateMipmaps for everything else. That way we don't break existing games, and support iphone in (what is hopefully) the usual case.
As an aside, webgl 2.0 has GL_TEXTURE_MAX_LEVEL support for using textures with partial mip chains.

Reply:
But as we've seen, the latest iPhones do not allow you to call ```generateMipmaps``` and then override certain mip levels. Currently, I'm feeling that the best path forwards is to say that user supplied mips must provide all levels or the texture is simply invalid. To fix After the Flood, I just had to add the missing 1x1 level. To handle legacy prefiltered cubemap DDSs, I would modify the cubemap resource loader (DDS reading code) and if there are missing levels, I would add them there. I don't think device.js should be hacking around this problem.